text/0000-dotdot-in-patterns.md

```diff
@@ -0,0 +1,129 @@
+
+# Detailed design
+[design]: #detailed-design
+
+Let's list all the patterns currently existing in the language, that contain lists of subpatterns:
+
+```
+// Struct patterns.
+S { field1, field2, ..., fieldN }
+
+// Tuple struct patterns.
+S(field1, field2, ..., fieldN)
+
+// Tuple patterns.
+(field1, field2, ..., fieldN)
+
+// Slice patterns.
+[elem1, elem2, ..., elemN]
+```
+In all the patterns above, except for struct patterns, field/element positions are significant.
+
+Now list all the contexts that currently permit the `..` pattern fragment:
+```
+// Struct patterns, the last position.
+S { subpat1, subpat2, .. }
+
+// Tuple struct patterns, the last and the only position, no extra subpatterns allowed.
+S(..)
+
+// Slice patterns, the last position.
+[subpat1, subpat2, ..]
+// Slice patterns, the first position.
+[.., subpatN-1, subpatN]
+// Slice patterns, any other position.
+[subpat1, .., subpatN]
+// Slice patterns, any of the above with a subslice binding.
+// (The binding is not actually a binding, but one more pattern bound to the sublist, but this is
+// not important for our discussion.)
+[subpat1, binding.., subpatN]
+```
+Something is obviously missing, let's fill in the missing parts.
+
+```
+// Struct patterns, the last position.
+S { subpat1, subpat2, .. }
+// **NOT PROPOSED**: Struct patterns, any position.
+// Since named struct fields are not positional, there's essentially no sense in placing the `..`
+// anywhere except for one conventionally chosen position (the last one) or in sublist bindings,
+// so we don't propose extensions to struct patterns.
+S { subpat1, .., subpatN }
+S { subpat1, binding.., subpatN }
+
+// Tuple struct patterns, the last and the only position, no extra subpatterns allowed.
+S(..)
+// **NEW**: Tuple struct patterns, any position.
+S(subpat1, subpat2, ..)
+S(.., subpatN-1, subpatN)
+S(subpat1, .., subpatN)
+// **NEW**: Tuple struct patterns, any position with a sublist binding.
+// The binding has a tuple type.
+// By ref bindings are not allowed, because layouts of S(A, B, C, D) and (B, C) are not necessarily
+// compatible (e.g. field reordering is possible).
+S(subpat1, binding.., subpatN)
+
+// **NEW**: Tuple patterns, any position.
+(subpat1, subpat2, ..)
+(.., subpatN-1, subpatN)
+(subpat1, .., subpatN)
+// **NEW**: Tuple patterns, any position with a sublist binding.
+// The binding has a tuple type.
+// By ref bindings are not allowed, because layouts of (A, B, C, D) and (B, C) are not necessarily
+// compatible (e.g. field reordering is possible).
+(subpat1, binding.., subpatN)
+
+// Slice patterns, the last position.
+[subpat1, subpat2, ..]
+// Slice patterns, the first position.
+[.., subpatN-1, subpatN]
+// Slice patterns, any other position.
+[subpat1, .., subpatN]
+// Slice patterns, any of the above with a subslice binding.
+// By ref bindings are allowed, slices and subslices always have compatible layouts.
+[subpat1, binding.., subpatN]
```

Comment:
>What does this binding mean?

A subslice pattern, see above - "The binding is not actually a binding, but one more pattern ..."